notebook/tests/selenium/test_display_id.py

```diff
@@ -0,0 +1,160 @@
+    def _get_iopub_messages(cell_idx):
+        script = f"return IPython.notebook.get_cell({cell_idx}).iopub_messages"
+        return notebook.browser.execute_script(script)
+    return _get_iopub_messages
+
+
+def test_display_id(prefill_notebook, get_outputs, get_iopub_messages):
+    notebook = prefill_notebook(INITIAL_CELLS)
+    notebook.execute_cell(0)
+
+    #TODO describe test 1
+    notebook.add_and_execute_cell(content=textwrap.dedent("""\
+        display('above')
+        display_with_id(1, 'here')
+        display('below')
+        """))
+    notebook.wait_for_cell_output(1)
+
+    cell1_outputs = get_outputs(1)
+    assert len(cell1_outputs) == 3, 'cell 1 has the right number of outputs'
+    assert cell1_outputs[1]['transient']['display_id'] == 'here', 'has transient display_id'
+    assert cell1_outputs[1]['data']['text/plain'] == '1', 'display_with_id produces output'
+
+    #TODO describe test 2
+    notebook.add_and_execute_cell(content=textwrap.dedent("""\
+        display_with_id(2, 'here')
+        display_with_id(3, 'there')
+        display_with_id(4, 'here')
+        """))
+    notebook.wait_for_cell_output(2)
+
+    cell1_outputs = get_outputs(1)
+    assert cell1_outputs[1]['data']['text/plain'] == '4'
+    assert len(cell1_outputs) == 3, 'cell 1 still has the right number of outputs'
+    cell2_outputs = get_outputs(2)
+    assert len(cell2_outputs) == 3, 'cell 2 has the right number of outputs'
+    assert cell2_outputs[0]['transient']['display_id'] == 'here', 'check display id 0'
+    assert cell2_outputs[0]['data']['text/plain'] == '4', 'output[2][0]'
+    assert cell2_outputs[1]['transient']['display_id'] == 'there', 'display id 1'
+    assert cell2_outputs[1]['data']['text/plain'] == '3', 'output[2][1]'
+    assert cell2_outputs[2]['transient']['display_id'] == 'here', 'display id 2'
+    assert cell2_outputs[2]['data']['text/plain'] == '4', 'output[2][2]'
+
+    # Test output callback overrides work with display ids.
+    notebook.browser.execute_script(textwrap.dedent("""\
+        Jupyter.notebook.insert_cell_at_index("code", 3);
+        var cell = Jupyter.notebook.get_cell(3);
+        cell.set_text([
+            "display_with_id(5, 'here')",
+            "display_with_id(6, 'here', update=True)",
+        ].join('\\n'));
+        cell.execute();
+
+        var cell = IPython.notebook.get_cell(3);
+        var kernel = IPython.notebook.kernel;
+        var msg_id = cell.last_msg_id;
+        var callback_id = 'mycallbackid'
+        cell.iopub_messages = [];
+        var add_msg = function(msg) {
+            msg.content.output_type = msg.msg_type;
+            cell.iopub_messages.push(msg.content);
+        };
+        kernel.set_callbacks_for_msg(callback_id, {
+            iopub: {
+                output: add_msg,
+                clear_output: add_msg,
+            }
+        }, false);
+        kernel.output_callback_overrides_push(msg_id, callback_id);
+        """))
```

Comment:
I ended up keeping most of this test in JavaScript. Why?

- Starting with the second half (L. 92-108), seems to me like these callbacks have to be set from JS, but this is the first time I've come across iopub, so maybe I'm missing something?
- As for the first bit (L. 84-90) - adding a cell and setting its contents - this could be done in selenium, via the UI. However, when I tried that, I started getting intermittent test failures. I played around with this in the browser and it turns out that adding just a slight delay (tens of milliseconds) on L. 91 causes the cell's `iopub_messages` array to remain empty.

It should be obvious from my comments that I don't know what I'm doing or what's going here! :)
If there's a better way to approach this or some documentation I could be reading to understand it better, I'd be all ears.